File: Othello/Project5/Othello.rb
#
# Othello Game Class
# Author(s): Jake Irons
#

class Othello

    # Constants
    WHITE = 'W'
    BLACK = 'B'
    EMPTY = '-'
    TIE = 'T'

    # Creates getter methods for instance variables @size, @turn, @disc,
    # @p1Disc, and @p2Disc
    attr_reader  :size, :turn, :disc, :p1Disc, :p2Disc

    # Constructs and initializes the board of given size
    def initialize(size, startPlayer, discColor)
        # validate arguments
        if size < 4 || size > 8 || size % 2 != 0
            raise ArgumentError.new('Invalid value for board size.')
        end
        if startPlayer < 1 || startPlayer > 2
            raise ArgumentError.new('Invalid value for player number.')
        end
        if discColor != WHITE && discColor != BLACK
            raise ArgumentError.new('Invalid value for disc.');
        end

        # set instance variables
        @board = []
        @size = size
        @turn = startPlayer
        @disc = discColor

        # set two more instance variables @p1Disc and @p2Disc
        if @turn == 1
            @p1Disc = @disc
            @p2Disc = @disc == WHITE ? BLACK : WHITE
        else
            @p2Disc = @disc
            @p1Disc = @disc == WHITE ? BLACK : WHITE;
        end

        # create the grid (as array of arrays)
        @board = Array.new(@size)
        for i in (0...@board.length)
            @board[i] = Array.new(@size)
            @board[i].fill(0)
        end

        # initialize the grid
        initializeBoard()
    end

    # Initializes the board with start configuration of discs
    def initializeBoard()
	for i in (0...@size)
		for j in (0...@size)
			@board[i][j] = '-'
		end
	end
	
	@board[@size/2 -1][@size/2 - 1] = 'B'
	@board[@size/2][@size/2] = 'B'
	@board[@size/2 - 1][@size/2] = 'W'
	@board[@size/2][@size/2 - 1] = 'W'	

    end

    # Returns true if placing the disc of current player at row,col is valid;
    # else returns false
    def isValidMove(row, col)
        return isValidMoveForDisc(row, col, @disc)
    end

    # Returns true if placing the specified disc at row,col is valid;
    # else returns false
    def isValidMoveForDisc(row, col, disc)
	other_disc = ''
	if disc == 'B'
		other_disc = 'W'
	else
		other_disc = 'B'
	end
	
	if @board[row][col] == '-'
		for surrRow in -1..1 do
			for surrCol in -1..1 do
				if surrRow != 0 or surrCol != 0
					new_row = row + surrRow
					new_col = col + surrCol
					if (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == other_disc
						while (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == other_disc do
							new_row = new_row + surrRow
							new_col = new_col + surrCol
							if (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == disc
								return true
							end
						end
					end
				end
			end
		end
	end
	
    	return false
    end

    # Places the disc of current player at row,col and flips the
    # opponent discs as needed
    def placeDiscAt(row, col)
        if (!isValidMove(row, col))
            return
        end

        # place the current player's disc at row,col
        @board[row][col] = @disc
	other_disc = ''
	if disc == 'B'
		other_disc = 'W'
	else
		other_disc = 'B'
	end
	
	for surrRow in -1..1 do
		for surrCol in -1..1 do
			if surrRow != 0 or surrCol != 0
				new_row = row + surrRow
				new_col = col + surrCol
				if (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == other_disc
					square_count = 1
					while (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == other_disc do
						square_count += 1
						new_row = new_row + surrRow
						new_col = new_col + surrCol
						if (new_row >= 0 and new_row < @size) and (new_col >= 0 and new_col < @size) and @board[new_row][new_col] == disc
							end_count = square_count - 1
							until end_count < 1 do
								@board[new_row - (surrRow * end_count)][new_col - (surrCol * end_count)] = @disc
								end_count -=  1
							end
							break
						end
					end
				end
			end
		end
	end
        

        # DO NOT DELETE - prepares for next turn if game is not over
        if (!isGameOver())
            prepareNextTurn();
        end
    end

    # Sets @turn and @disc instance variables for next player
    def prepareNextTurn()
        if @turn == 1
            @turn = 2
        else
            @turn = 1
        end
        if @disc == WHITE
            @disc = BLACK
        else
            @disc = WHITE
        end
    end

    # Returns true if a valid move for current player is available;
    # else returns false
    def isValidMoveAvailable()
        isValidMoveAvailableForDisc(@disc)
    end

    # Returns true if a valid move for the specified disc is available;
    # else returns false
    def isValidMoveAvailableForDisc(disc)

        for i in 0...@size do
        	for j in 0...@size do
        		if isValidMoveForDisc(i, j, disc)
        			return true
        		end
        	end
        end

        # DO NOT DELETE - if control reaches this statement, then a valid move is not available
        return false;
    end

    # Returns true if the board is fully occupied with discs; else returns false
    def isBoardFull()

        for i in 0...@size do
        	for j in 0...@size do
        		if @board[i][j] == '-'
        			return false
        		end
        	end
        end

        return true
    end

    # Returns true if either the board is full or a valid move is not available
    # for either disc
    def isGameOver()
        return isBoardFull() ||
                    (!isValidMoveAvailableForDisc(WHITE) &&
                                !isValidMoveAvailableForDisc(BLACK))
    end

    # If there is a winner, it returns Othello::WHITE or Othello::BLACK.
    # In case of a tie, it returns Othello::TIE
    def checkWinner()

        black_count = 0
        white_count = 0
        if !isGameOver
        	return
        else
        	for i in 0...@size do
        		for j in 0...@size do
        			if @board[i][j] == 'B'
        				black_count += 1
        			elsif @board[i][j] == 'W'
        				white_count += 1
        			end
        		end
        	end
        end
        
        if white_count > black_count
        	return Othello::WHITE
        elsif black_count > white_count
        	return Othello::BLACK
        end
        
        return Othello::TIE
    end

    # Returns a string representation of the board
    def to_s()
        str = "\n  "
        for i in (0...@size)
            str << (i+1).to_s + ' '
        end
        str << "\n";
        for i in (0...@size)
            str << (i+1).to_s + ' ';
            str << @board[i].join(' ') + "\n";
        end
        return str;
    end

end

File: Othello/Project3/Othello.js
'use strict';

/**
 * Othello game
 * Author(s): Jake Irons
 */

class Othello {

	// Constructs and initializes the board of given size
	constructor(size, startPlayer, discColor) {
		// validate arguments
		if (size < 4 || size > 8 || size % 2 !== 0) {
			throw new Error("Invalid value for board size.");
		}
		if (startPlayer < 1 || startPlayer > 2) {
			throw new Error("Invalid value for player number.");
		}
		if (discColor !== Othello.WHITE && discColor !== Othello.BLACK) {
			throw new Error("Invalid value for disc.");
		}

		// set instance variables
		this.size = size;
		this.turn = startPlayer;
		this.disc = discColor;

		// set two more instance variables p1Disc and p2Disc
		if (this.turn === 1) {
			this.p1Disc = this.disc;
			this.p2Disc = this.disc === Othello.WHITE ? Othello.BLACK : Othello.WHITE;
		} else {
			this.p2Disc = this.disc;
			this.p1Disc = this.disc === Othello.WHITE ? Othello.BLACK : Othello.WHITE;
		}

		// create the grid (as array of arrays)
		this.board = new Array(this.size);
		for (let i = 0; i < this.board.length; i++) {
			this.board[i] = new Array(this.size);
			this.board[i].fill(0);
		}

		// initialize the grid
		this.initializeBoard();
	}

	// Getter for white disc
  	static get WHITE() {
    	return "W";
  	}

  	// Getter for black disc
  	static get BLACK() {
    	return "B";
  	}

  	// Getter for empty
  	static get EMPTY() {
    	return "-";
  	}

  	// Getter for tie
  	static get TIE() {
    	return "T";
  	}

  	// Initializes the board with start configuration of discs (as per project specs)
	initializeBoard() {
		for(let i = 0; i < this.board.length; i++){
			for(let j = 0; j < this.board.length; j++){
				this.board[i][j] = '-';
			}
		}
		this.board[this.size / 2 - 1][this.size / 2 - 1] = 'B';
		this.board[this.size / 2][this.size / 2] = 'B';
		this.board[this.size / 2 - 1][this.size / 2] = 'W';
		this.board[this.size / 2][this.size / 2 - 1] = 'W';
  	}

	// Returns true if placing the disc of current player at row,col is valid; else returns false
	isValidMove(row, col) {
		return this.isValidMoveForDisc(row, col, this.disc);
	}

	// Returns true if placing the specified disc at row,col is valid; else returns false
	isValidMoveForDisc(row, col, disc) {
		var otherDisc;
		if(disc == 'B'){
			otherDisc = 'W';
		}
		else{
			otherDisc = 'B';
		}

		if(this.board[row][col] == '-'){
			var newRow;
			var newCol;
			for(var surrRow = -1; surrRow < 2; surrRow++){
				for(var surrCol = -1; surrCol < 2; surrCol++){
					if(surrRow != 0 || surrCol != 0){
						newRow = row + surrRow;
						newCol = col + surrCol;
						if((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length)){
							if(this.board[newRow][newCol] == otherDisc){
								while((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length) && this.board[newRow][newCol] == otherDisc) {
			    						newRow = newRow + surrRow;
									newCol = newCol + surrCol;
									if((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length)){
										if(this.board[newRow][newCol] == disc){
											return true;
										}
									}
								}
							}
						}
					}
				}
			}
		}


		// DO NOT DELETE - if control reaches this statement, then it is not a valid move
		return false;	// not a valid move
	}

	// Places the disc of current player at row,col and flips the opponent discs as needed
	placeDiscAt(row, col) {
		if (!this.isValidMove(row, col)) {
			return;
		}

		// place the current player's disc at row,col
		this.board[row][col] = this.disc;
		var otherDisc;
		if(this.disc == 'B'){
			otherDisc = 'W';
		}
		else{
			otherDisc = 'B';
		}
		var newRow;
		var newCol;
		for(let surrRow = -1; surrRow < 2; surrRow++){
			for(let surrCol = -1; surrCol < 2; surrCol++){
				if(surrRow != 0 || surrCol != 0){
					newRow = row + surrRow;
					newCol = col + surrCol;
					if((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length)){
						if(this.board[newRow][newCol] == otherDisc){
							var squareCount = 1;
							while((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length) && this.board[newRow][newCol] == otherDisc) {
								++squareCount;
		    						newRow = newRow + surrRow;
								newCol = newCol + surrCol;
								if((newRow >=0 && newRow < this.board.length) && (newCol >= 0 && newCol < this.board.length)){
									if(this.board[newRow][newCol] == this.disc){
										for(var i = squareCount - 1; i > 0; i--){
											this.board[newRow - (surrRow * i)][newCol - (surrCol * i)] = this.disc;
										}
										break;
									}
								}
							
							}
						}
					} 
				}
			}
		}



		// DO NOT DELETE - prepares for next turn if game is not over
		if (!this.isGameOver()) {
			this.prepareNextTurn();
		}
	}

	// Sets turn and disc information for next player
	prepareNextTurn() {
		if (this.turn === 1) {
			this.turn = 2;
		} else {
			this.turn = 1;
		}
		if (this.disc === Othello.WHITE) {
			this.disc = Othello.BLACK;
		} else {
			this.disc = Othello.WHITE;
		}
	}

	// Returns true if a valid move for current player is available; else returns false
	isValidMoveAvailable() {
		return this.isValidMoveAvailableForDisc(this.disc);
	}

	// Returns true if a valid move for the specified disc is available; else returns false
	isValidMoveAvailableForDisc(disc) {

		for (let i = 0; i < this.board.length; i++) {
			for (let j = 0; j < this.board.length; j++) {
				if(this.isValidMove(i, j, disc)){
					return true;
				}
			}
		}

		
		// DO NOT DELETE - if control reaches this statement, then a valid move is not available
		return false;	// not a valid move
	}

	// Returns true if the board is fully occupied with discs; else returns false
	isBoardFull() {

		for (let i = 0; i < this.board.length; i++) {
			for (let j = 0; j < this.board.length; j++) {
				if(this.board[i][j] == '-'){
					return false;
				}
			}
		}

		return true;
	}

	// Returns true if either the board is full or a valid move is not available for either disc
	isGameOver() {
		return this.isBoardFull() ||
					(!this.isValidMoveAvailableForDisc(Othello.WHITE) &&
								!this.isValidMoveAvailableForDisc(Othello.BLACK));
	}

	// If there is a winner, it returns Othello.WHITE or Othello.BLACK.
	// In case of a tie, it returns Othello.TIE
	checkWinner() {

		var blackCount = 0; 
		var whiteCount = 0;
		if (!this.isGameOver()) {
			return;
		}
		else{
			for (var i = 0; i < this.board.length; i++) {
				for (var j = 0; j < this.board.length; j++) {
					if(this.board[i][j] == 'B'){
						++blackCount;
					}
					else if(this.board[i][j] == 'W'){
						++whiteCount;
					}
				}
			}
		}
		if(whiteCount > blackCount){
			return Othello.WHITE;
		}
		else if(blackCount > whiteCount){
			return Othello.BLACK;
		}
		return Othello.TIE;
		
	}

	// Returns a string representation of the board (for display purposes)
	toString() {
		let str = '\n ';
		for (let i = 0; i < this.size; i++) {
			str += ' ' + (i+1)
		}
		str += "\n";
		for (let i = 0; i < this.size; i++) {
			str += i+1 + ' ';
			str += this.board[i].join(' ') + "\n";
		}
		return str;
	}
}

module.exports = Othello;
